### DISCUSSION TITLE
Typescript 5?

### DISCUSSION THREAD

**Discussion by JaneJeon** [mikro-orm/mikro-orm#3974] [Ideas]
Just wanted to throw this out there to start a discussion, not asking for anything specific.

MikroORM is very heavily reliant on decorators, and we currently use experimentalDecorators flag to enable those.

However, upon reading the TS 5@beta release notes, I've been made aware that "the type-checking rules and emit are sufficiently different that while decorators _can_ be written to support both the old and new decorators' behavior, any existing decorator functions are not likely to do so."

As a TS newbie, I'm not quite sure what this means for MikroORM - will it "just work" with TypeScript v5, or will we be bound to v4 with experimentalDecorators and emitDecoratorMetadata?

**Comment by B4nan** [comment]
If you want to use property type inference, which is the main reason for decorators, you will need `emitDecoratorMetadata`, hence the legacy decorators. They explicitly say this will be supported for foreseeable future.

You can already use ES decorators, e.g. people trying out esbuild are already using this specification - note that you won't get any type inferred, so providing it explicitly is required:

```diff
-@Property({ nullable: true })
+@Property({ type: 'string', nullable: true }) // or type: t.string
foo?: string;
```

As far as I know, the decorator definition we have should be compatible already. I will wait at least for the release candidate to try it out.

**Reply by JaneJeon** [reply]
Thanks!
